### ISSUE TITLE
Portrait vs landscape modes when using android camera2

### ISSUE THREAD

**Issue by Tim-Kindberg** [pedroSG94/rtmp-rtsp-stream-client-java#599]
I'd like for the streamed video to be portrait when I hold the phone in portrait mode, and landscape when I hold the phone in landscape mode. I don't need this behaviour to be dynamic - i.e. it's only the orientation when the user presses the "go live" button that counts.

I'm using android's camera2 and a surface exported from a VideoEncoder to live-stream. I notice that the Texture/Surface mode RTMP examples are doing the same thing. 

The phone's orientation (rotation) is passed to VideoEncoder.prepareVideoEncoder, and the dimensions given to srsFlvMuxer.setVideoResolution are swapped according to the state of ratation, but this doesn't seem to have any effect: the streamed video is always in landscape mode. This is true of the Texture/Surface mode RTMP examples as well. 

BTW, I'm setting hardwareRotation = false - I'm not sure what that does (or whether it's available across android devices).

How do I rotate the streamed video to match how the user is holding the phone?

Many thanks,

Tim

**Comment by pedroSG94** [comment]
Hello,
If you are using camera2 and TextureView or SurfaceView you have a problem because you need rotate image before render your surface.
You have multiples ways to do it:

- Rotate with encoder. You can set rotation to encoder but not all encoder support it (most don't have it):
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/video/VideoEncoder.java#L119
- Rotate with camera. I don't know how to implement it with camera2 but with camera1 you can select orientation used to render a surface like this:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/video/Camera1ApiManager.java#L181
- Rotate with opengl. This is maybe the best way and I use it with OpenglView the idea is rotate image obtained from camera before render your surface (in this case videoEncoder surface) and you can modify it as you want on fly.

In my case, hardwareRotation in videoEncoder indicate that you will rotate with encoder or not.

If you select true, the videoEncoder rotate surface and image for you. For example, if you select 1280x720 and you set rotation true of 90º you have a surface of 720x1280 and original image rotated 90º.
If you select false I'm doing a resolution swap if you set 90º or 270º and image rotation is done by opengl:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/video/VideoEncoder.java#L104

**Comment by pedroSG94** [comment]
Yes, if you rotate with my first suggestion this will affect to all surfaces set to your camera implementation.
About OpenGlView. This is really difficult to explain because opengl is difficult to understand. Maybe the best idea is use LightOpenGlView because it is only a render without filters so it is more simple.

You need study this classes:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/view/LightOpenGlView.java
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/SimpleCameraRender.java
Basically surface created in SimpleCameraRender should be used to render with your camera implenentation (this will draw in your view) and you should set your VideoEncoder surface to LightOpenGlView after init it:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/view/OpenGlViewBase.java#L109
Matrix in SimpleCameraRender is used for rotation and fragment.glsl if you want modify image (filters, etc)

**Comment by Tim-Kindberg** [comment]
Thanks Pedro,

I've tried a simple Experiment without rotation but I get an exception. Apparently I'm not initialising something correctly but I've tried to follow what you do in your own code. The exception is:

>     java.lang.IllegalArgumentException: Make sure the SurfaceView or associated SurfaceHolder has a valid Surface
>         at android.opengl.EGL14._eglCreateWindowSurface(Native Method)
>         at android.opengl.EGL14.eglCreateWindowSurface(EGL14.java:265)
>         at com.pedro.encoder.input.gl.SurfaceManager.eglSetup(SurfaceManager.java:129)
>         at com.pedro.encoder.input.gl.SurfaceManager.<init>(SurfaceManager.java:46)
>         at com.pedro.rtplibrary.view.LightOpenGlView.run(LightOpenGlView.java:78)
>         at java.lang.Thread.run(Thread.java:919)
> 

My code is:

```
// I use the Activity as Context 
mLightOpenGlView = new LightOpenGlView( mActivity );
mLightOpenGlView.init();
// Take no account of rotation for now
mLightOpenGlView.setEncoderSize( mVideoEncoder.getWidth(), mVideoEncoder.getHeight() );
// No rotation for now
mLightOpenGlView.setRotation( 0 );
mLightOpenGlView.start();
mLightOpenGlView.addMediaCodecSurface( mVideoEncoder.getInputSurface() );

```
It seems to be failing in start() - before I've added the VideoEncoder's Surface, but that is the correct place to call addMediaCodecSurface() according to your code. 

Which "SurfaceView or associated SurfaceHolder" is it referring to?

**Comment by pedroSG94** [comment]
Maybe LightOpenGlView surface is not created yet. Try add this callback to know when you can call start method:
```java
mLightOpenGlView.getHolder().addCallback(new SurfaceHolder.Callback() {
      @Override
      public void surfaceCreated(SurfaceHolder surfaceHolder) {
        
      }

      @Override
      public void surfaceChanged(SurfaceHolder surfaceHolder, int i, int i1, int i2) {
        //surface is ready you can use start method
      }

      @Override
      public void surfaceDestroyed(SurfaceHolder surfaceHolder) {
        //surface released. Stop method is called automatically
      }
    });
```

**Comment by Tim-Kindberg** [comment]
Thanks Pedro. I added that callback immediately after creation of the LightOpenGlView - which happens some time before mLightOpenGlView.init() even though I showed them together in the code snippet above. But none of the surfaceCreated() etc. methods is ever called - after init, setEncoderSize, setRotation.

I frankly don't understand what the SurfaceHolder returned by getHolder() is. I Notice that in OpenGlRtmpActivity you have:

`openGlView = findViewById(R.id.surfaceView);`

In other words you seem to be constructing it from a resource, not from scratch, as I have.

I think I must be missing something altogether in my initialisation.

**Comment by pedroSG94** [comment]
Basically I added that to my layout.xml (this view is drawn in my screen).
If your methods isn't called is because you are not drawing it and you need do it.

Other way could be use OffScreenGlThread class instead of LightOpenGlView if you don't want a view.

**Comment by Tim-Kindberg** [comment]
Thanks again for all your help, Pedro. That worked (sort of). In portrait mode, I now see the image in Facebook Live the right way up. However, it is not complete and letterboxed, as I was expecting: it's on the left hand side, cropped  somewhat vertically, and possibly a little stretched horizontally. Is that perhaps something to do with my setup? The image is 480x720. 

I was expecting to have to set the rotation on the OffScreenGlThread to 90/270 but the image came the right way up with that set to zero.

**Comment by pedroSG94** [comment]
Maybe you need modify matrix and viewport to adapt image:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/utils/gl/SizeCalculator.java
Your viewport should be always this line (this should be encoderSize if you use OffScreenGlThread):
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/utils/gl/SizeCalculator.java#L36
And mode is always 0 so this should be the only line executed if you don't set flip:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/utils/gl/SizeCalculator.java#L55

**Comment by Tim-Kindberg** [comment]
Thanks but the only use I see of SizeCalculator is in a couple of classes, e.g. the drawFrame method of SimpleCameraRender - which I'm not using since I'm using OffScreenGlThread. In SimpleCameraRender it's used in a sequence of OpenGl and matrix operations. It's not at all clear to me how I would apply SizeCalculator in OffScreenGlThread, where there are no matrix operations at all. What am I missing?

**Comment by pedroSG94** [comment]
That class is used in all opengl classes (OpenGlView, LightOpenGlView and OffScreenGlThread). This is the route to SizeCalculator with OffScreenGlThread:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/view/OffScreenGlThread.java#L219

https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/ManagerRender.java#L66

https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/gl/render/ScreenRender.java#L84

viewport is pixel range where you start draw and you stop draw. 
For example, I have a surface of 100x100 to draw if you want draw all surface you should start at pixel 0x0 and finish in pixel 100x100 (glViewport(0, 0, previewWidth, previewHeight)).

About matrix you can control scale (scale range is 0 to 1 where 1 is full viewport size) and rotation of the image.

**Comment by Tim-Kindberg** [comment]
Hi Pedro, yes I saw that SizeCalculator is eventually called in my case. I feel like I must be missing something completely from what you are suggesting, since I still have no idea what I would be able to change from a software engineering point of view:
1. All the parameters to SizeCalculator are completely determined, ultimately, by the parameters I provide to OffScreenGlThread - I could only change the OffScreenGlThread parameters, in particular from the desired 480x720 to something artificially modified to get the right result - which would be what?
2. I can't (also) meaningfully call SizeCalculator outside the existing flow of logic inside your classes. I'd have to modify (in this case) ScreenRender.draw() if SizeCalculator was to be called differently.
3. I'm not setting flip, and I'm providing the desired (correct) width and height. So from your comments above I don't know what I'd do differently when calling SizeCalculator in ScreenRender.draw() anyway.

I don't think you're suggesting (1). Are you suggesting (2)? Probably not. What else is there?!

**Comment by pedroSG94** [comment]
Basically you need modify the class (SizeCalculator) to work as you want. For example, you only want draw half surface modify this line:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/utils/gl/SizeCalculator.java#L36
To:
```java
GLES20.glViewport(0, 0, previewWidth / 2, previewHeight / 2);
```
Similar if you want [rotate](https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/utils/gl/SizeCalculator.java#L61) or move image to avoid crops. Maybe I'm doing a confuse explain for it.
Can you share me an image with your problem and other image with the result expected?

**Comment by Tim-Kindberg** [comment]
Hi Pedro,

Screenshot of app on phone showing camera preview (480x720 letterboxed):

![Screenshot_20200618-174558 copy](https://user-images.githubusercontent.com/1196660/85050877-b1855880-b18e-11ea-8905-94c008cf9f06.png)


Screenshot of Facebook Live preview on laptop:

![live](https://user-images.githubusercontent.com/1196660/85050767-7be06f80-b18e-11ea-8107-e92fec165066.png)

It seems obvious, now that I'm using the newspaper image, that it's a 480x480 square crop that's being rendered - and it doesn't seem like it's being stretched in either direction. I want to see the whole 480x720 image in the centre of the Facebook Live frame - which is what I get when I use your app.

So I want to draw more of the image (all of it), and move it to a central position. 

Thanks.

**Comment by pedroSG94** [comment]
This seems that your camera resolution and video resolution is different. At least in images aspect ratio is different. Maybe Your prepareVideo in VideoEncoder is the problem.
Also We can check that full image is send by camera doing a scale of image to avoid crops. MOdify this line:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/utils/gl/SizeCalculator.java#L60
To:
```java
Matrix.scaleM(MVPMatrix, 0, scale.x * 0.5f, scale.y * 0.5f, 1f);
```

**Comment by Tim-Kindberg** [comment]
Hi Pedro,

Thanks. I solved my problem with a hack to SizeCalculator.calculateViewPort - as you were suggesting.  You have a similar calculation to mine but in the branch:
`if( keepAspectRatio) {`
But keepAspectRatio is false. Also, my previewWidth and previewHeight are constants determined by the Camera2 setup, and always previewWidth > previewHeight even if the user holds the phone in portrait mode.

**Comment by pedroSG94** [comment]
In my case keepAspectRatio is only for preview and with a parameter of view enabled but OffScreenGlThread is background (no preview in this case) for that reason always is false.

I glad that all is working now :D
